### PR TITLE
Example and test of custom messages

### PR DESCRIPTION
**Pull request by cragun47**
Signed-off-by: Brian Cragun <cragun@us.ibm.com>

This shows how a custom message can be created and used with the converter.
The test ensures that we don't break this capability in the future.

### CODE REVIEW

**Comment by cragun47** [comment]
tag @LisaWellman

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
+/*
```
Comment (LisaWellman) [inline src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:1]: Suggestion: renames CUSTOM_PATIENT just to be a little clearer

File: src/test/java/custom_packages/2.6 [hunk 2]
```diff
@@ -0,0 +1 @@
+org.foo.hl7.custom
```
Comment (LisaWellman) [inline src/test/java/custom_packages/2.6:1]: Can we add a comment in this file explaining what this is for / what it has to match?

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+ * <ul>
+ * <li>MSH (Message Header) <b> </b></li>
+ * <li>PID (Patient Identification) <b> </b></li>
+ * <li>PRB (Problem Details) <b>optional repeating</b></li>
+ * <li>AL1 (Allergy) <b>optional repeating</b></li>
+ * </ul>
+ *
+ * Sample Message:
+ * MSH|^~\\&|||||20210709162149||HIST^PAT|20210709162149|P|2.6
+ * PID|1|100021^^^FAC|100021^^^FAC||DOE^JANE||196705270000|F|||||5733551967|||U ++
```
Comment (LisaWellman) [inline src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:30]: Extraneous `++` at end of line here and line 32

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+ * PRB|1|20210726|11334^ABNORMALITIES OF
+ * HAIR^ICD9||||201208070948||201208070948|||||||201208070948 ++
```
Comment (LisaWellman) [inline src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:32]: This should be the same line as previous (not 2 lines)

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+			this.add(PID.class, true, false);
+			this.add(PRB.class, false, true);
+			this.add(AL1.class, false, true);
+		} catch (HL7Exception e) {
+			log.error("Unexpected error creating CUSTOM_PAT - this is probably a bug in the source code generator.");
+		}
+	}
+
+	/**
+	 * Returns "2.6"
```
Comment (LisaWellman) [inline src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:66]: Remove actual version from the comment, just say "version"

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+	 */
+	public String getVersion() {
+		return "2.6";
+	}
+
+	/**
+	 * <p>
+	 * Returns MSH (Message Header) - creates it if necessary
+	 * </p>
+	 *
```
Comment (LisaWellman) [inline src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:76]: Remove empty lines in comments - general comment

I would also remove the paragraph tags as they add no value.

File: src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+# (C) Copyright IBM Corp. 2021
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# FHIR Resources to extract from CUSTOM_PAT message
+#
+
+########################################################################
+# Used for testing only. Not used in production.
+########################################################################
```
Comment (LisaWellman) [inline src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml:11]: I like this block - can we add to CUSTOM_PAT.java as well

File: src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+---
+resources:
+    - resourceName: MessageHeader
+      segment: MSH
+      resourcePath: resource/MessageHeader
+      repeats: false
+      isReferenced: false
+      additionalSegments: 
+        - EVN
```
Comment (LisaWellman) [inline src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml:21]: There is no EVN segment in this message, remove additionalSegments

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
+package io.github.linuxforhealth.hl7.message;
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java:1]: Move below copyright

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+// http://javadox.com/ca.uhn.hapi/hapi-base/2.1/ca/uhn/hl7v2/parser/DefaultModelClassFactory.html#packageList(java.lang.String)
+// In this test, the custom class which HL7 uses for validation is placed in src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java
+// The custom message is placed in src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml
+// The custom packages class is placed in src/test/java/custom_packages/2.6 and references the custom package /org/foo/hl7/custom/
+
+public class Hl7CustomMessageTest {
+
+    // # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+    // NOTE VALIDATION IS INTENTIONALLY TURNED OFF BECAUSE WE ARE CREATING RESOURCES THAT ARE NOT STANDARD
+    // private static final ConverterOptions OPTIONS = new Builder().withValidateResource().withPrettyPrint().build()
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java:49]: Remove commented out line, the comment above covers the point

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+    @Test
+    public void testCustomPatMessage() throws IOException {
+
+        // Set up the config file
+    	commonConfigFileSetup();
+
+        ClassLoader loader = Test.class.getClassLoader();
+        System.out.println(loader.getResource("custom_packages/2.6"));
+        System.out.println(loader.getResource("org/foo/hl7/custom/message/CUSTOM_PAT.class"));
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java:90]: Remove println

**Review by LisaWellman — APPROVED** [review]
Small cleanup items, looks good!

File: src/test/java/custom_packages/2.6 (hunk 2 above)
Comment (cragun47) [reply src/test/java/custom_packages/2.6:1]: Done.

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java (hunk 4 above)
Comment (cragun47) [reply src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:32]: Done.  (Darn formatter).

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java (hunk 3 above)
Comment (cragun47) [reply src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:30]: removed.

File: src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java (hunk 6 above)
Comment (cragun47) [reply src/test/java/org/foo/hl7/custom/message/CUSTOM_PAT.java:76]: Done

File: src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml (hunk 7 above)
Comment (cragun47) [reply src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml:11]: Done

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java (hunk 9 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java:1]: done

File: src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml (hunk 8 above)
Comment (cragun47) [reply src/test/resources/additional_custom_resources/hl7/message/CUSTOM_PAT.yml:21]: Done

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java (hunk 11 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java:90]: Done.

File: src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java (hunk 10 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/message/Hl7CustomMessageTest.java]: Done

**Pull request merged**